### ISSUE TITLE
Command to Comment/Uncomment

### ISSUE THREAD

**Issue by JoelProminic** [prominic/Moonshine-IDE#575]
One common IDE feature that is still missing is a command to comment/uncomment a line or selected block of code.  I have generally seen this bound to CMD-/ and CTRL-/.   This is not an urgent change, so it can be included in a release if someone is waiting for other tasks to be completed.

For Java, Groovy, and Actionscript, insert "//" at the very beginning (column 0) of the line.   I prefer this method to inserting just before the text because it keeps the code clean if there are indents.

When uncommenting, this should only affect comment characters at the beginning of the line, ignoring whitepace.

If the selection has mixed commented and uncommented lines, comment all of the lines again.  This will result in nested commenting, but it preserves existing comments when/if the code is uncommented.

Don't use the language block comments (i.e "/* ... */"), because:
- This requires more complicated cases when there are embedded block comments already
- Commenting the lines individually allows them to be uncommented individually.

For XML, HTML, and MXML (also XHTML and GSP, if we add specific support for them), we need to use HTML-style comments instead ().   This case is a little tricker, but my specific recommendations are:
- Apply the comment to each line individually
- Put the opening comment text at the beginning of the line - this is something IntelliJ doesn't do, but I think it will look better like this
- If there are comments inside the commented line/selection, they need to be escaped to avoid breaking the code.  I would suggest "&lt;!--" and "--&gt;", but I am open to better ideas.

These are the two main file types that I expect us to support, but we can additional logic for other file types as needed.  

In the long term, we could also consider supporting server-side comments (i.e. <%-- --%>) for GSP and XHTML files.  Here are some [examples](https://examples.javacodegeeks.com/enterprise-java/jsp/comments-in-jsp-page/).

**Comment by piotrzarzycki21** [comment]
@joshtynjala This issue is related to editor. Do you want me to create issue in [moonshine-feathersui-text-editor](https://github.com/Moonshine-IDE/moonshine-feathersui-text-editor/) ?

**Comment by joshtynjala** [comment]
@piotrzarzycki21 Keyboard shortcuts to toggle comments are already implemented in moonshine-feathersui-text-editor, and I just confirmed that they are working in Moonshine IDE. Ctrl+/ toggles a line comment for the selected line(s), and Ctrl+Alt+/ toggles a block comment for the selected range.

I don't think that menu items have been added to Moonshine yet. However, if anyone wants to add that, the `TextEditor` class also has public `toggleLineComment()` and `toggleBlockComment()` methods that may be called.

**Comment by JoelProminic** [comment]
I did some tests with Java projects.   The line comments work mostly like I expect. I see that the comments follow the indents currently:

![image](https://user-images.githubusercontent.com/24532731/198338661-2b5563bf-7513-4dc8-86b5-331875c87a5c.png)

The editors I normally use put the comments at the beginning of the line:
- Advantage:  this makes it clear that this is a block of code that was commented out together
- Disadvantage:  the `//` is at an unexpected position for nested code, especial when it is deep nesting.   
![image](https://user-images.githubusercontent.com/24532731/198338930-550644f9-8056-4f57-8fa8-6fb70d0aac5f.png)

This may come down to developer preference, so we can keep the current behavior if this is what others prefer.

**Comment by JoelProminic** [comment]
I generally don't use block comments (i.e. `/* ... */` for commenting out blocks of code, because it breaks if there is a nested block comment.

The toggle works well for me for a single line:
![image](https://user-images.githubusercontent.com/24532731/198344071-fdcdd74b-5f83-47af-986c-17d828a29e65.png)

However, I found that the position of the `/*` and `*/` depends on the direction that I drag the mouse.  If I start at the end of the line, I get:
![image](https://user-images.githubusercontent.com/24532731/198343985-50f0b67c-cb11-44f2-9b19-4bc407d78c44.png)


For multiple lines, the selectionupdates to include the trailing comment.  If I immediately try to uncomment this block, it will instead add another block comment.

![image](https://user-images.githubusercontent.com/24532731/198343948-47af0d21-6b39-4fd0-840d-030d03c51392.png)

If the selection ends at the trailing `}` instead, the block uncomments properly.

**Comment by joshtynjala** [comment]
> I see that the comments follow the indents currently

Good catch. I actually intended to match how VSCode does it, where it follows only the shortest indent among the selected lines. More like this:

```java
public class TestJavaGradle1027
{
	// public static void main(String args[])
	// {
	// 	// Existing comment
	// 	System.out.println("Hello World!");
	// }
}
```

I'll try to get that fixed when I have a chance.

> I found that the position of the /* and */ depends on the direction that I drag the mouse.
> If I immediately try to uncomment this block, it will instead add another block comment.

Definitely bugs I'll need to get fixed.

**Comment by JoelProminic** [comment]
@solaechea and @dxhereira reported that `CTRL-/` didn't work on Windows.  I don't remember which projects they were testing.

I did a quick test with a Java Gradle project on Windows, and `CTRL-/` worked for me.

**Comment by solaechea** [comment]
> @solaechea and @dxhereira reported that `CTRL-/` didn't work on Windows. I don't remember which projects they were testing.
> 
> I did a quick test with a Java Gradle project on Windows, and `CTRL-/` worked for me.

I was testing with this Java class: https://github.com/Moonshine-IDE/FormBuilder-Contact-API/blob/main/ContactsAgents/src/main/java/ContactAgents/ContactCreate.java

After some testing I think that it fails because my keyboard is an spanish keyboard, the layout is different as you can see in the next image:
![spanish keyboard](https://user-images.githubusercontent.com/1082804/203789954-9f2da28b-37d1-4503-8220-a82ad3820580.png)

### Using Spanish or English keyboard layout
The forward slash in the numeric pad (number 3 in the image above) doesn't work for comment shortcut.

### Using Spanish layout
To get forward slash I need to type `shift + 7` (number 1 in the image above), so `CTRL + shift + 7 ` doesn't work for comment shortcut.

### Using English layout (Spanish keyboard but switching keyboard layout in Windows settings)
The shortcur works with `CTRL + the key number 2 in the image above`.

I think is is similar to https://forum.sublimetext.com/t/many-shortcuts-fail-on-non-us-keyboard/15712

**Comment by joshtynjala** [comment]
Moonshine-IDE/moonshine-feathersui-text-editor@237926132afb336eea12c6a48a4c81c0c307fccd
Moonshine-IDE/moonshine-feathersui-text-editor@78929ebcfc7c79f2b6ab8f5535c055fd7b736bb9
Moonshine-IDE/moonshine-feathersui-text-editor@64e3fadedd7606c121e0324cc9aa30793c9786ac

**Comment by solaechea** [comment]
I Tested this on Windows 11 with a Spanish keyboard and I can toggle comments (single and multiple lines) with the numpad slash key. I tested a java, xml and js files.
